File: app/src/main/java/com/krystianwsul/checkme/firebase/records/RemoteRootUserRecord.java
package com.krystianwsul.checkme.firebase.records;

import android.support.annotation.NonNull;
import android.text.TextUtils;

import com.krystianwsul.checkme.firebase.UserData;
import com.krystianwsul.checkme.firebase.json.UserJson;
import com.krystianwsul.checkme.firebase.json.UserWrapper;

import junit.framework.Assert;

import java.util.Map;

public class RemoteRootUserRecord extends RemoteRecord {
    private static final String USER_DATA = "userData";
    private static final String FRIEND_OF = "friendOf";

    @NonNull
    private final UserWrapper mUserWrapper;

    public RemoteRootUserRecord(boolean create, @NonNull UserWrapper userWrapper) {
        super(create);

        mUserWrapper = userWrapper;
    }

    @NonNull
    public UserJson getUserJson() {
        return mUserWrapper.getUserData();
    }

    @NonNull
    @Override
    protected UserWrapper getCreateObject() {
        return mUserWrapper;
    }

    @NonNull
    public String getId() {
        return UserData.getKey(getUserJson().getEmail());
    }

    @NonNull
    @Override
    protected String getKey() {
        return getId();
    }

    @NonNull
    public String getName() {
        return getUserJson().getName();
    }

    @NonNull
    public String getEmail() {
        return getUserJson().getEmail();
    }

    public void setName(@NonNull String name) {
        if (getName().equals(name))
            return;

        getUserJson().setName(name);
        addValue(getKey() + "/" + USER_DATA + "/name", name);
    }

    public void removeFriendOf(@NonNull String friendId) {
        Assert.assertTrue(!TextUtils.isEmpty(friendId));

        Map<String, Boolean> friendOf = mUserWrapper.getFriendOf();
        Assert.assertTrue(friendOf != null);
        Assert.assertTrue(friendOf.containsKey(friendId));

        friendOf.remove(friendId);

        addValue(getKey() + "/" + FRIEND_OF + "/" + friendId, null);
    }
}

File: app/src/main/java/com/krystianwsul/checkme/firebase/records/RemoteRootUserRecord.kt
package com.krystianwsul.checkme.firebase.records

import com.krystianwsul.common.firebase.UserData
import com.krystianwsul.common.firebase.json.UserWrapper


open class RemoteRootUserRecord(create: Boolean, override val createObject: UserWrapper) : RemoteRecord(create) {

    companion object {

        const val USER_DATA = "userData"
        private const val FRIEND_OF = "friendOf"
    }

    val userJson by lazy { createObject.userData }

    val id by lazy { UserData.getKey(userJson.email) }

    override val key by lazy { id }

    var name: String
        get() = userJson.name
        set(name) {
            if (name == userJson.name)
                return

            userJson.name = name
            addValue("$key/$USER_DATA/name", name)
        }

    val email by lazy { userJson.email }

    open val photoUrl get() = userJson.photoUrl

    fun removeFriendOf(friendId: String) {
        check(friendId.isNotEmpty())

        val friendOf = createObject.friendOf
        check(friendOf.containsKey(friendId))
        checkNotNull(friendOf[friendId])

        friendOf.remove(friendId)

        addValue("$key/$FRIEND_OF/$friendId", null)
    }

    override fun deleteFromParent() = throw UnsupportedOperationException()
}
